src/goto-instrument/cover.cpp

```diff
@@ -88,11 +121,117 @@ class basic_blockst
     block_line_cover_mapt;
   block_line_cover_mapt block_line_cover_map;
 
-  inline unsigned operator[](goto_programt::const_targett t)
+  /// \param t a goto instruction
+  /// \return the block number of the block
+  ///         the given goto instruction is part of
+  unsigned block_of(goto_programt::const_targett t)
   {
     return block_map[t];
   }
 
+  /// \param block_nr a block number
+  /// \return the instruction that has been selected for
+  ///   instrumentation representative of the given block
+  goto_programt::const_targett instruction_of(
+    unsigned block_nr)
+  {
+    return block_loc[block_nr];
+  }
+
+  /// Select an instruction to be instrumented for each basic block such that
+  /// the java bytecode indices for each basic block is unique
+  /// \param goto_program The goto program
+  void select_unique_java_bytecode_indices(const goto_programt &goto_program)
+  {
+    std::set<unsigned> blocks_seen;
+    std::set<irep_idt> bytecode_indices_seen;
+
+    forall_goto_program_instructions(it, goto_program)
+    {
+      unsigned block_nr=block_map[it];
+      if(blocks_seen.find(block_nr)!=blocks_seen.end())
+        continue;
+
+      goto_programt::const_targett in_t=block_loc[block_nr];
+      if(in_t==goto_program.instructions.end())
+      {
+        if(!it->source_location.get_java_bytecode_index().empty())
+        {
+          // search for a representative
+          if(bytecode_indices_seen.find(
+               it->source_location.get_java_bytecode_index())==
+             bytecode_indices_seen.end())
+          {
+            block_loc[block_nr]=it;
+            irep_idt lines_covered=
+              source_location_map[block_nr].get_basic_block_covered_lines();
+            source_location_map[block_nr]=it->source_location;
+            source_location_map[block_nr]
+              .set_basic_block_covered_lines(lines_covered);
+            bytecode_indices_seen.insert(
+              it->source_location.get_java_bytecode_index());
+            blocks_seen.insert(block_nr);
+          }
+        }
+      }
+      else if(it==in_t)
+      {
+        // check the existing representative
```

Comment:
Suggest `// check the existing representative instruction doesn't have the same bytecode index as a block we've already seen`